### PR TITLE
[Feature] Add RewardModelLoss objective for RLHF reward-model training

### PR DESCRIPTION
**Pull request by theap06**
## What

Adds a model-agnostic Bradley-Terry pairwise **reward-model loss** as a first-class objective under `torchrl/objectives/llm/`, following the `SFTLoss` / `GRPOLoss` conventions (`_AcceptedKeys` / `set_keys`, a `TensorClass` output, runnable docstring with paper references).

## Why

The Bradley-Terry reward loss already existed as `GPT2RewardModel.compute_reward_loss` (`torchrl/modules/models/llm.py`), explicitly tagged `# TODO: move to objectives` and tied to GPT2. This promotes it to a proper objective that:

- lives in `objectives/` like every other loss;
- is **model-agnostic** — it owns a swappable `score_network` (any `TensorDictModule` producing a per-sequence scalar, e.g. `AutoModelForSequenceClassification(num_labels=1)` or `GPT2RewardModel`);
- is TensorDict-first with configurable keys.

## Contents

- `torchrl/objectives/llm/reward.py`: `RewardModelLoss`, `RewardModelLossOutput`, and the `reward_model_loss` helper. Supports `mean`/`sum`/`none` reduction, an optional score-centering regularizer, and a detached `accuracy` metric for logging.
- Exports from `torchrl/objectives/llm/__init__.py`.
- Docs: "Reward Model Training" sections in `llms_objectives.rst` and `llms.rst`.
- Tests: `TestRewardModel` in `test/llm/test_llm_objectives.py` (CPU-only, includes a nested-key input case per the contributor guide).

## Tests

`pytest test/llm/test_llm_objectives.py -k RewardModel` → 9 passed. `ruff` + `black` clean; the public-API `__all__` guard passes.

### CODE REVIEW

File: torchrl/objectives/llm/reward.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+                if self.device is not None
+                else contextlib.nullcontext()
+            ):
+                sub_td = self.score_network(sub_td)
+        score = sub_td.get(self.tensor_keys.score, default=None)
+        if score is None:
+            raise KeyError(
+                f"Could not find the score at key {self.tensor_keys.score!r} under "
+                f"{key!r}. If score_network is None, the scores must be precomputed."
+            )
```
Comment (vmoens) [inline torchrl/objectives/llm/reward.py:263]: after set_keys(score=("metrics", "score")), _score invokes an unchanged score_network and then reads only the customized key. A network still writing its declared/default "score" key breaks. The nested-key test covers only score_network=None.

File: torchrl/objectives/llm/reward.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        *,
+        reduction: Literal["mean", "sum", "none"] = "mean",
+        center_coeff: float | None = None,
+        device: torch.device | str | None = None,
+    ) -> None:
+        super().__init__()
+        self.score_network = score_network
+        self.reduction = reduction
+        self.center_coeff = center_coeff
+        self.device = torch.device(device) if device is not None else None
```
Comment (vmoens) [inline torchrl/objectives/llm/reward.py]: self.device is a plain attribute, so loss.to(...) does not keep it synchronized

In general we should avoid to have modules that have a self.device attribute.

File: torchrl/objectives/llm/reward.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            )
+        if self.device is not None:
+            sub_td = sub_td.to(self.device)
+        if self.score_network is not None:
+            with (
+                torch.device(self.device)
+                if self.device is not None
+                else contextlib.nullcontext()
+            ):
+                sub_td = self.score_network(sub_td)
```
Comment (vmoens) [inline torchrl/objectives/llm/reward.py]: ditto, self.device should not really be a thing IMO. Let's think about models that have more than one device for example.

**Review by vmoens — COMMENTED** [review]
Two additional fixes are needed beyond those that I posted:
- invalid reduction values are rejected only after both score-network passes
- the example scorer averages padding tokens and does not demonstrate the advertised attention_mask contract

Thanks for all the contributions!

**Comment by theap06** [comment]
@vmoens thanks for the feedback! Just addressed it. 
image

File: torchrl/objectives/llm/reward.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    def _score(self, tensordict: TensorDictBase, key: NestedKey) -> torch.Tensor:
+        """Run the score network on a sub-tensordict and return a per-sequence score."""
+        sub_td = tensordict.get(key, default=None)
+        if sub_td is None:
+            raise KeyError(
+                f"Could not find the sub-tensordict at key {key!r} in the input "
+                f"tensordict with keys {set(tensordict.keys())}."
+            )
+        if self.score_network is not None:
+            sub_td = self.score_network(sub_td)
```
Comment (gtnv) [inline torchrl/objectives/llm/reward.py]: this scorer writes score into the caller-owned chosen and rejected TensorDicts. can we run it on a shallow structural clone?

```suggestion
            sub_td = self.score_network(sub_td.clone(False))
```

File: torchrl/objectives/llm/reward.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        score_network: TensorDictModule | None = None,
+        *,
+        reduction: Literal["mean", "sum", "none"] = "mean",
+        center_coeff: float | None = None,
+    ) -> None:
+        _validate_reduction(reduction)
+        super().__init__()
+        self.score_network = score_network
+        self.reduction = reduction
+        self.center_coeff = center_coeff
```
Comment (gtnv) [inline torchrl/objectives/llm/reward.py:226]: center_coeff=-1.0 is accepted rn. that makes the squared-score term lower the objective. can we reject negative values?

```suggestion
        if center_coeff is not None and center_coeff < 0:
            raise ValueError("center_coeff must be non-negative.")
        self.center_coeff = center_coeff
```

File: torchrl/objectives/llm/reward.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+_REDUCTIONS = ("mean", "sum", "none")
+
+
+def _validate_reduction(reduction: str) -> None:
+    if reduction not in _REDUCTIONS:
+        raise ValueError(f"Invalid reduction: {reduction}.")
+
+
+def reward_model_loss(
```
Comment (vmoens) [inline torchrl/objectives/llm/reward.py]: We should bundle all the functional like this one in a functionals.py file
